pennylane/decomposition/resources.py

```diff
@@ -0,0 +1,347 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the data structure that stores resource estimates for each decomposition."""
+
+from __future__ import annotations
+
+import functools
+from dataclasses import dataclass, field
+from functools import cached_property
+from typing import Optional, Type
+
+import pennylane as qml
+from pennylane.operation import Operator
+
+
+@dataclass(frozen=True)
+class Resources:
+    r"""Stores resource estimates.
+
+    Args:
+        gate_counts (dict): dictionary mapping operator types to their number of occurrences.
+
+    """
+
+    gate_counts: dict[CompressedResourceOp, int] = field(default_factory=dict)
+
+    def __post_init__(self):
+        """Verify that the gate counts and the number of gates are consistent."""
+        assert all(v > 0 for v in self.gate_counts.values())
+
+    @cached_property
+    def num_gates(self) -> int:
+        """The total number of gates."""
+        return sum(self.gate_counts.values())
+
+    def __add__(self, other: Resources):
+        return Resources(
+            _combine_dict(self.gate_counts, other.gate_counts),
+        )
+
+    def __mul__(self, scalar: int):
+        return Resources(_scale_dict(self.gate_counts, scalar))
+
+    __rmul__ = __mul__
+
+    def __repr__(self):
+        return f"num_gates={self.num_gates}, gate_counts={self.gate_counts}"
+
+
+def _combine_dict(dict1: dict, dict2: dict):
+    r"""Combines two dictionaries and adds values of common keys."""
+
+    combined_dict = dict1.copy()
+
+    for k, v in dict2.items():
+        combined_dict[k] = combined_dict.get(k, 0) + v
+
+    return combined_dict
+
+
+def _scale_dict(dict1: dict, scalar: int):
+    r"""Scales the values in a dictionary with a scalar."""
+
+    scaled_dict = dict1.copy()
+
+    for k in scaled_dict:
+        scaled_dict[k] *= scalar
+
+    return scaled_dict
```

Comment:
```suggestion
    return {key: scalar*value for key, value in dict1.items()}
```
Only one pass the size of the dictionary this way, instead of two. Also, now it's just one line, so we might not even need to extract it.